### DISCUSSION TITLE
How to synchronize `puts` output?

### DISCUSSION THREAD

**Discussion by postmodern** [socketry/async#227] [Q&A] [answered]
I've noticed that if I have a lot of async sub-tasks which pass data to a shared callback, which then prints the data using `puts`, sometimes the newlines are doubled up as if Ruby tries to write both values at once then write the newline characters.

## Example

```ruby
require 'async'

def test(&block)
  Async do |task|
    task.async do
      loop do
        block.call("foooooooooooooooooooooooo")
        sleep(0.1)
      end
    end

    task.async do
      loop do
        block.call("baaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaar")
        sleep(0.1)
      end
    end
  end
end

test do |word|
  puts word
end
```

```
$ ruby test_printing.rb 
foooooooooooooooooooooooobaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaar

foooooooooooooooooooooooo
baaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaar
foooooooooooooooooooooooo
baaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaar
foooooooooooooooooooooooo
baaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaar
foooooooooooooooooooooooo
baaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaar
foooooooooooooooooooooooo
```

How do I prevent this from happening?

**Comment by ioquatix** [comment, marked as answer]
IIRC, this should be fixed in the latest version of CRuby. What version did you test on?

**Reply by postmodern** [reply]
Can confirm this doesn't happen under `ruby-3.2.0`, but does happen under `ruby-3.1.4`.

**Reply by ioquatix** [reply]
Awesome, yeah there were some odd code in `io.c` w.r.t. locking. It was too fine grained, so because `puts` is actually two writes, the string itself followed by a newline, before the lock was taken per write, now it's taken for the entire puts.

Here are the commits that made more operations atomic: https://github.com/ruby/ruby/commits/6b0131db6f029f3e472d00deb2134d45ad4dde3a
